addons/metallb/0.9.x/metallb-1.yaml

```diff
@@ -16,6 +16,15 @@ spec:
   cloudProvider:
     - name: azure
       enabled: false
+    - name: vsphere
+      enabled: true
+      values: |
+        configInline:
+          address-pools:
+          - name: default
+            protocol: layer2
+            # configure addresses for your network
+            addresses: []
```

Comment:
This effectively means MetalLB will be broken (by default) on vsphere.

Can we provide a default IP range that will be used and available via the VSphere network deployment for our Konvoy VSphere provisioner?

Reply:
How should this work?
We don't know anything about customers IP ranges.
We need to document that there must be an IP range added to that list.

Reply:
With the docker provisioner we use the default docker network range as that will be available for host networking by default.

With the VSphere provisioner is a default range for the host network not known?

Reply:
I expect user setups the whole vSphere setting on its own. So also DNS, DHCP or not DHCP (so static IPs), Timeserver, etc pp.
There is basically no default.

Reply:
Given that as the case my suggestion would be that we take this approach for now to keep current progress unblocked:

1. we remove the enabling of metallb for vpshere for now, as it can not be automatically managed as a default in current state
2. we document heavily for this iteration that manual intervention is needed to fully deploy the addons (e.g. metallb needs to be manually configured and enabled, and instructions on how to do that)
3. we create a JIRA which covers the technical need to add fully automated `LoadBalancer` svcs in the case of the VSphere provisioner
4. we make the JIRA from the above a blocker for VSphere provisioner release

Also we would need to get product team buy-in on this approach, cc @samba and @GoelDeepak.

Let me know your thoughts on this approach @sebbrandt87? :thinking:

Reply:
(NOTE: also started a [Slack Thread](https://mesosphere.slack.com/archives/CMG6Y3C9X/p1599150474006500))

Reply:
@sebbrandt87 @shaneutt  I reviewed with @GoelDeepak and we've come to these changes...

* Use NodePort by default for Traefik in onprem environments & vsphere provider
* Document configuring Traefik to use LoadBalancer if available in onprem and vsphere environments
* Disable metallb by default onprem and vsphere providers
* Document the correct use of metallb to provide Service[type=LoadBalancer] in each environment
* Konvoy should look for the keepalived address and nodeport allocated to the Service for Traefik, when onprem/vsphere, instead of the current LoadBalancer external address, when printing opsportal access information

The main problems are these:
* We need sensible defaults that work out-of-the-box with zero configuration for at least a limited set of use-cases.
* We need to provide usable templates with documentation, for alternative configurations.
* We've seen numerous customer cases where third-party onprem clusters do not support Service[type=LoadBalancer], and need to accommodate this better in our addons stack.
* Konvoy will consider the cluster to be _fully functional_ only when Traefik becomes operational.

Unfortunately there's no environment where metallb can work out-of-the-box with any appropriate defaults; therefore it will require manual configuration, and therefore needs documentation. It also should therefore be disabled by default.  (I realize this is a change of direction from past ideas, but these use-cases are clarifying the needed behavior.)

We even explored the terraform vsphere provider to determine if there were good ways to pre-allocate some service IPs for metallb... and found nothing. So that's a non-starter. vSphere simply doesn't operate at the same level as other cloud providers, in respect to managed load balancers.

Naturally we'd welcome any additional feedback or alternative ideas.

Reply:
My only feedback is that I'm generally in alignment with the proposed changes, however for this wording:

> Unfortunately there's no environment where metallb can work out-of-the-box with any appropriate defaults;

I would caveat that this is _currently_ true, but that we could look into future engineering efforts to improve our MetalLB offering in this regard which is what we started discussing in Slack.

@sebbrandt87 for your purposes to keep you moving forward: I'm glad we had this conversation but if you want to simply remove the metallb default deployment from vsphere I will consider this comment resolved and be ready to :heavy_check_mark: the PR.

Reply:
non-blocking: just wanted to make sure I followed up with a supplemental KEP @dkoshkin and I authored which could help us improve things to better enable end-users to use MetalLB (and not have `NodePort` by default as their only option):

https://github.com/mesosphere/ksphere-platform/pull/207

Reply:
> > Unfortunately there's no environment where metallb can work out-of-the-box with any appropriate defaults;
> 
> I would caveat that this is currently true, but that we could look into future engineering efforts to improve our MetalLB offering in this regard which is what we started discussing in Slack.

This will always be true. There's no way to predict or interpret the desired network configuration where we don't control the network. Take an example where there are 4 network interfaces, a default on 10.1.0.0/16, a backend network on 10.2.0.0/16, a frontend network of 10.3.0.0/16, and the gateway network on 192.168.0.0/24. There's no way to interpret which network the customer wants to allocate addresses from.

Reply:
> > > Unfortunately there's no environment where metallb can work out-of-the-box with any appropriate defaults;
> > 
> > 
> > I would caveat that this is currently true, but that we could look into future engineering efforts to improve our MetalLB offering in this regard which is what we started discussing in Slack.
> 
> This will always be true. There's no way to predict or interpret the desired network configuration where we don't control the network. Take an example where there are 4 network interfaces, a default on 10.1.0.0/16, a backend network on 10.2.0.0/16, a frontend network of 10.3.0.0/16, and the gateway network on 192.168.0.0/24. There's no way to interpret which network the customer wants to allocate addresses from.

Practically what you're saying is correct, but I don't see any argument against this: see the KEP which asks for improved flagging for the Konvoy CLI:

https://github.com/mesosphere/ksphere-platform/pull/207